pkg/util/span/frontier.go

```diff
@@ -185,229 +276,283 @@ func (f *Frontier) PeekFrontierSpan() roachpb.Span {
+//
+// NB: it is unsafe for the caller to modify the keys in the provided span after this
+// call returns.
+func (f *btreeFrontier) Forward(
+	span roachpb.Span, ts hlc.Timestamp,
+) (forwarded bool, retErr error) {
+	// Validate caller provided span.
+	if err := checkSpan(span); err != nil {
 		return false, err
 	}
-	return prevFrontier.Less(f.frontierLocked()), nil
-}
-
-// extendRangeToTheLeft extends the range to the left of the range, provided those
-// ranges all have specified timestamp.
-// Updates provided range with the new starting position.
-// Returns the list of frontier entries covered by the updated range; the caller
-// is expected to remove those covered ranges from the tree.
-func extendRangeToTheLeft(
-	t interval.Tree, r *interval.Range, ts hlc.Timestamp,
-) (covered []*frontierEntry) {
-	for {
-		// Get the range to the left of the range.
-		// Since we request an inclusive overlap of the range containing exactly
-		// 1 key, we expect to get two extensions if there is anything to the left:
-		// the range (r) itself, and the one to the left of r.
-		left := t.GetWithOverlapper(
-			interval.Range{Start: r.Start, End: r.Start},
-			interval.InclusiveOverlapper,
-		)
-		if len(left) == 2 && left[0].(*frontierEntry).ts.Equal(ts) {
-			e := left[0].(*frontierEntry)
-			covered = append(covered, e)
-			r.Start = e.keys.Start
-		} else {
-			return
-		}
+
+	if expensiveChecksEnabled() {
+		defer func() {
+			if err := f.checkUnsafeKeyModification(); err != nil {
+				retErr = errors.CombineErrors(retErr, err)
+			}
+		}()
 	}
-}
 
-// extendRangeToTheRight extends the range to the right of the range, provided those
-// ranges all have specified timestamp.
-// Updates provided range with the new ending position.
-// Returns the list of frontier entries covered by the updated range; the caller
-// is expected to remove those covered ranges from the tree.
-func extendRangeToTheRight(
-	t interval.Tree, r *interval.Range, ts hlc.Timestamp,
-) (covered []*frontierEntry) {
-	for {
-		// Get the range to the right of the range.
-		// Since we request an exclusive overlap of the range containing exactly
-		// 1 key, we expect to get exactly 1 extensions if there is anything to the right of the span.
-		endKey := roachpb.Key(r.End)
-		rightSpan := roachpb.Span{Key: endKey, EndKey: endKey.Next()}
-		right := t.GetWithOverlapper(rightSpan.AsRange(), interval.ExclusiveOverlapper)
-		if len(right) == 1 && right[0].(*frontierEntry).ts.Equal(ts) {
-			e := right[0].(*frontierEntry)
-			covered = append(covered, e)
-			r.End = e.keys.End
-		} else {
-			return
-		}
+	prevFrontier := f.Frontier()
+	if err := f.forward(span, ts); err != nil {
+		return false, err
 	}
+	return prevFrontier.Less(f.Frontier()), nil
 }
 
-func (f *Frontier) insert(span roachpb.Span, insertTS hlc.Timestamp) error {
-	// Set of frontier entries to add and remove.
-	var toAdd, toRemove []*frontierEntry
+// clone augments generated iterStack code to support cloning.
```

Comment:
nit: perhaps open an issue or leave a todo in the btree template to add this function?